### PR TITLE
avalon-zmq

### PR DESCRIPTION
**Pull request by divyataori**
Avalon integration with ZMQ messaging library for synchronous work order processing and a new sample application for cold chain evaluation.

### CODE REVIEW

File: tc/sgx/trusted_worker_manager/enclave/CMakeLists.txt [hunk 1]
```diff
@@ -127,7 +127,7 @@ ENDFUNCTION()
 
 # XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
 
-FILE(GLOB PROJECT_HEADERS *.h ${TCF_CORE_DIR}/common/*.h)
+FILE(GLOB PROJECT_HEADERS *.h)
```
Comment (manju956) [inline tc/sgx/trusted_worker_manager/enclave/CMakeLists.txt]: These changes to Cmake file is no more needed as the latest code already have it covered by PR - https://github.com/hyperledger/avalon/pull/235

**Comment by manju956** [comment]
> Avalon integration with ZMQ messaging library for synchronous work order processing and a new sample application for cold chain evaluation.

Commit message doesn't have Signed-Off field, add it.

**Comment by pankajgoyal2** [comment]
DCO is failing...please commit with Sign-Off

File: client_sdk/avalon_client_sdk/quorum/contracts/WorkOrder.sol [hunk 2]
```diff
@@ -0,0 +1,172 @@
+pragma solidity >=0.4.0;
+
```
Comment (Ram-srini) [inline client_sdk/avalon_client_sdk/quorum/contracts/WorkOrder.sol]: We already have workorder contract in ethereum folder, can't we use that contract for quorum?

File: client_sdk/avalon_client_sdk/quorum/quorum_worker_registry_list_client.py [hunk 3]
```diff
@@ -0,0 +1,233 @@
+# Copyright 2019 Intel Corporation
```
Comment (Ram-srini) [inline client_sdk/avalon_client_sdk/quorum/quorum_worker_registry_list_client.py]: We are in the process of creating one SDK for avalon. We are already has PR#234 for that.

File: client_sdk/avalon_client_sdk/tcf_connector.toml [hunk 4]
```diff
@@ -24,30 +24,69 @@ json_rpc_uri = "http://localhost:1947"
...
-eth_http_provider = "https://ropsten.infura.io/v3/9af17e6320414af8a3168cd9fe9d653d"
+eth_http_provider = "https://ropsten.infura.io/"
 #chain_id is 3 for ropsten test network
 #"1": Ethereum Mainnet
 #"2": Morden Testnet (deprecated)
 #"3": Ropsten Testnet
 #"4": Rinkeby Testnet
 #"42": Kovan Testnet
-chain_id = 3
+chain_id = 27609
```
Comment (Ram-srini) [inline client_sdk/avalon_client_sdk/tcf_connector.toml]: What is this chain id corresponds to?

File: examples/apps/cold_chain_eval/client/ColdChainAPI.py [hunk 5]
```diff
@@ -0,0 +1,392 @@
+# Client for Coldchain workload
+import os
+from os import urandom, path, environ
+from flask import request
+from flask import Flask
+from avalon_client_sdk.quorum.quorum_worker_registry_client import QuorumWorkerRegistryImpl
```
Comment (danintel) [inline examples/apps/cold_chain_eval/client/ColdChainAPI.py]: You need to run `bin/run_lint` and fix any formatting issues in this and other files.

File: docker/Dockerfile.tcf-dev [hunk 6]
```diff
@@ -196,7 +196,7 @@ ENV DISPLAY=:0
 
 WORKDIR /project/TrustedComputeFramework/tools/build
 
-RUN make lint && echo "LINT SUCCESS"
+# RUN make lint && echo "LINT SUCCESS"
```
Comment (danintel) [inline docker/Dockerfile.tcf-dev]: Do not disable lint. Instead fix the warnings and errors.

File: docker-compose.yaml [hunk 7]
```diff
@@ -29,6 +29,10 @@ services:
         - TCF_DEBUG_BUILD=${TCF_DEBUG_BUILD:-}
         - DISPLAY=${DISPLAY:-}
         - XAUTHORITY=~/.Xauthority
+    ports:
+      - "1947:1947"
+      - "5000:5000"
```
Comment (danintel) [inline docker-compose.yaml]: What is port 5000 needed for?
I see port 5555 used for ZMQ, but no usage of port 5000 anywhere.

File: config/tcs_config.toml [hunk 8]
```diff
@@ -32,6 +32,8 @@ sleep_interval = "10"
 [Listener]
 listener_port = 1947
 max_work_order_count = 10
+zmq_port = '5555'
```
Comment (danintel) [inline config/tcs_config.toml]: Add a 1 line comment about zmq_port.

File: client_sdk/avalon_client_sdk/utility/hex_utils.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
+# Copyright 2019 Intel Corporation
```
Comment (danintel) [inline client_sdk/avalon_client_sdk/utility/hex_utils.py]: Add Apache 2.0 license.

File: client_sdk/avalon_client_sdk/quorum/contracts/WorkOrder.sol [hunk 10]
```diff
@@ -0,0 +1,172 @@
+pragma solidity >=0.4.0;
```
Comment (danintel) [inline client_sdk/avalon_client_sdk/quorum/contracts/WorkOrder.sol]: Add Apache 2.0 license.

File: client_sdk/avalon_client_sdk/quorum/__init__.py [hunk 11]
```diff
@@ -0,0 +1,2 @@
+
```
Comment (danintel) [inline client_sdk/avalon_client_sdk/quorum/__init__.py]: Add Apache 2.0 license.

File: client_sdk/avalon_client_sdk/interfaces/work_invocation_interface.py [hunk 12]
```diff
@@ -0,0 +1,55 @@
+from abc import ABC,abstractmethod
```
Comment (danintel) [inline client_sdk/avalon_client_sdk/interfaces/work_invocation_interface.py]: Add Apache 2.0 license.

**Comment by danintel** [comment]
You need to add a DCO line to your commit as explained here:
https://github.com/hyperledger/avalon/blob/master/CONTRIBUTING.md
This is required by The Linux Foundation.

**Comment by manojgop** [comment]
Please break it down into multiple smaller PRs. For example you can raise a separate PR for quorum integration, ZMQ messaging library for synchronous work order processing and cold chain sample application. This will help to review, rework and merge independent functionalities easily.

File: client_sdk/avalon_client_sdk/quorum/__init__.py (hunk 11 above)
Comment (divyataori) [reply client_sdk/avalon_client_sdk/quorum/__init__.py]: Updated in the latest commit.Please review

File: client_sdk/avalon_client_sdk/interfaces/work_invocation_interface.py (hunk 12 above)
Comment (divyataori) [reply client_sdk/avalon_client_sdk/interfaces/work_invocation_interface.py]: Updated in the latest commit.Please review

**Comment by divyataori** [comment]
> You need to add a DCO line to your commit as explained here:
> https://github.com/hyperledger/avalon/blob/master/CONTRIBUTING.md
> This is required by The Linux Foundation.

I have missed to add signoff in some last commits. Is there any workaround?

File: client_sdk/avalon_client_sdk/tcf_connector.toml (hunk 4 above)
Comment (divyataori) [reply client_sdk/avalon_client_sdk/tcf_connector.toml]: This chain_id corresponds to chain_id of private blockchain network. It is configured while the blockchain network is being setup.

File: examples/apps/cold_chain_eval/README.md [hunk 13]
```diff
@@ -0,0 +1 @@
+Update here
```
[the remark this answers is not in the document]
Comment (divyataori) [reply examples/apps/cold_chain_eval/README.md]: Updated in the latest commit.

**Comment by danintel** [comment]
You can do this for your last commit.
git commit -s –amend

Otherwise, for the last 3 commits (for example), type this:

git rebase -i HEAD~3

Edit the list of commits as instructed in the file (change “pick” to “reword”) and add the DCO signature line.

Force push your commit to your branch on github.com. For example, if the branch was named “zmq”, type this:

git push -f origin zmq

Dan

File: docker-compose.yaml (hunk 7 above)
Comment (danintel) [reply docker-compose.yaml]: Remove this line. TCP Port is not used by Avalon or ZMQ.

Comment (divyataori) [reply docker-compose.yaml]: The 5000 port is used by the Colchain client flask server. File location -examples/apps/cold_chain_eval/client/ColdChainAPI.py

Comment (danintel) [reply docker-compose.yaml]: OK. Add a 1 line comment for what ports 5000 and 555 are used for.

File: client_sdk/avalon_client_sdk/quorum/contracts/WorkOrder.sol (hunk 2 above)
Comment (divyataori) [reply client_sdk/avalon_client_sdk/quorum/contracts/WorkOrder.sol]: There are some modifications that are required by the Coldchain application.

File: client_sdk/avalon_client_sdk/quorum/quorum_worker_registry_client.py [hunk 14]
```diff
@@ -0,0 +1,258 @@
...
+    import WorkerRegistryClient
+from avalon_client_sdk.utility.utils import construct_message, validate_details
+
+logging.basicConfig(
+    format="%(asctime)s - %(levelname)s - %(message)s", level=logging.INFO)
+
+
+class QuorumWorkerRegistryImpl(WorkerRegistryClient):
+    """
+    This class is to read worker registries, it interract quorum blockchain.
```
[the remark this answers is not in the document]
Comment (divyataori) [reply client_sdk/avalon_client_sdk/quorum/quorum_worker_registry_client.py]: Changed in the latest commits.

File: client_sdk/avalon_client_sdk/quorum/contracts/WorkOrder.sol [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+        bytes32 indexed workerId,
+        bytes32 tag,
+        bytes32 nonce,
+        bytes signature,
+        uint32 errorCode
+    );
+
+    //-------------------------Function definitions----------------------------
+
+    //Retrict call from authorised requester
```
[the remark this answers is not in the document]
Comment (divyataori) [reply client_sdk/avalon_client_sdk/quorum/contracts/WorkOrder.sol]: Updated

**Comment by divyataori** [comment]
> Please break it down into multiple smaller PRs. For example you can raise a separate PR for quorum integration, ZMQ messaging library for synchronous work order processing and cold chain sample application. This will help to review, rework and merge independent functionalities easily.

Actually the Coldchain Client listens to the zmq socket communication from enclave. All changes are inter-dependent, so cannot separate into different PRs.

**Comment by manojgop** [comment]
> ZMQ messaging library for synchronous work order processing

 ZMQ messaging library for synchronous work order processing should be implemented a generic change which should be independent of any applications. Please raise a separate PR for this change.

**Pull request closed without merging**